sotodlib/site_pipeline/make_focal_plane.py

```diff
@@ -0,0 +1,264 @@
+
+        basis: Basis to calculate width in.
+               Currently not implemented so width will just be along the dets axis.
+               At the very least radial distance will be added.
+
+    Returns:
+
+        priors: The 2d array of priors.
+    """
+
+    def _flat(arr, idx):
+        arr[idx - width // 2 : idx + width // 2 + width % 2] = prior
+
+    if method == "flat":
+        prior_method = _flat
+    else:
+        raise ValueError("Method " + method + " not implemented")
+
+    priors = np.ones((aman.dets.count, aman.dets.count))
+    for i in aman.dets.count:
+        prior_method(priors[i], i)
+
+    return priors
+
+
+def match_template(
+    focal_plane, template, out_thresh=0, avoid_collision=True, priors=None
+):
+    """
+    Match fit focal plane againts a template.
+
+    Arguments:
+
+        focal_plane: Focal plane after either having optics model applied or being fit with channel map.
+                     Should have columns: x, y or x, y, pol.
+
+        template: Table containing template of focal plane.
+                  Should have columns: x, y or x, y, pol.
+
+        out_thresh: Threshold at which points will be considered outliers.
+                    Should be in range [0, 1) and is checked against the
+                    probability that a point matches its mapped point in the template.
+
+        avoid_collision: Try to avoid collisions. May effect performance.
+
+        priors: Priors to apply when matching.
+                Should be be a n by n array where n is the number of points.
+                The priors[i, j] is the prior on the i'th point in template matching the j'th point in focal_plane.
+
+    Returns:
+
+        mapping: Mapping between elements in template and focal_plane.
+                 focal_plane[i] = template[mapping[i]]
+
+        outliers: Indices of points that are outliers.
+                  Note that this is in the basis of mapping and focal_plane, not template.
+    """
+    reg = AffineRegistration(**{"X": focal_plane, "Y": template})
+    reg.register()
+    P = reg.P
+
+    if priors is not None:
+        P *= priors
+    if avoid_collision:
+        # This should get the maximum probability without collisions
+        inv = np.linalg.pinv(P)
+        mapping = np.argmax(inv, axis=0)
+    else:
+        mapping = np.argmax(P, axis=1)
+
+    outliers = np.array([])
+    if out_thresh > 0:
+        outliers = np.where(reg.P[range(reg.P.shape[0]), mapping] < out_thresh)[0]
+
+    return mapping, outliers
+
+
+def main():
+    # Read in input pars
+    parser = ap.ArgumentParser()
+
+    # NOTE: Eventually all of this should just be metadata I can load from a single context?
+
+    # Making some assumtions about pointing data that aren't currently true:
+    # 1. I am assuming that the HDF5 file is a saved aman not a pandas results set
+    # 2. I am assuming that it contains the measured detector polangs
+    # 3. I am assuming it comtains aman.det_info
```

Comment:
@Zhilei here are the interface assumptions I am making that are currently incorrect. I imagine we will land on something between what currently exists and what I am assuming here but I wanted to make the assumptions that required the least extra code here for the first pass.

Reply:
This makes sense, thanks for taking the first step with what we have!